encoding/marshal.go

```diff
@@ -94,111 +113,164 @@ func marshal(val reflect.Value) (b []byte) {
+	return b.Bytes()
+}
+
+// MarshalAll encodes all of its inputs and returns their concatenation.
+func MarshalAll(v ...interface{}) []byte {
+	b := new(bytes.Buffer)
+	enc := NewEncoder(b)
+	for i := range v {
+		enc.Encode(v[i]) // no error possible when using a bytes.Buffer
+	}
+	return b.Bytes()
+}
+
+// WriteFile writes v to a file. The file will be created if it does not exist.
+func WriteFile(filename string, v interface{}) error {
+	file, err := os.OpenFile(filename, os.O_WRONLY|os.O_CREATE, 0660)
+	if err != nil {
+		return err
+	}
+	return NewEncoder(file).Encode(v)
+}
+
+// A Decoder reads and decodes values from an input stream.
+type Decoder struct {
+	r io.Reader
 }
 
-// Unmarshal decodes a byte slice into the provided interface. The interface must be a pointer.
-// The decoding rules are the inverse of those described under Marshal.
-func Unmarshal(b []byte, v interface{}) (err error) {
+// Decode reads the next encoded value from its input stream and stores it in
+// v, which must be a pointer. The decoding rules are the inverse of those
+// specified in the package docstring.
+func (d *Decoder) Decode(v interface{}) (err error) {
 	// v must be a pointer
 	pval := reflect.ValueOf(v)
 	if pval.Kind() != reflect.Ptr || pval.IsNil() {
-		return errors.New("must pass a valid pointer to Unmarshal")
+		return errors.New("must pass a valid pointer to Decode")
 	}
 
-	// unmarshal may panic
-	// note that this allows us to skip any boundary checks while unmarshalling
-	var consumed int
+	// catch decoding panics and convert them to errors
+	// note that this allows us to skip boundary checks during decoding
 	defer func() {
-		if r := recover(); r != nil || consumed != len(b) {
-			err = errors.New("could not unmarshal type " + pval.Elem().Type().String())
+		if r := recover(); r != nil {
+			err = fmt.Errorf("could not decode type %s: %v", pval.Elem().Type().String(), r)
 		}
 	}()
 
-	consumed = unmarshal(b, pval.Elem())
+	d.decode(pval.Elem())
 	return
 }
 
-func unmarshal(b []byte, val reflect.Value) (consumed int) {
+func (d *Decoder) readN(n int) []byte {
+	b := make([]byte, n)
+	_, err := io.ReadFull(d.r, b)
+	if err != nil {
+		panic(err)
+	}
+	return b
+}
+
+func (d *Decoder) readPrefix() []byte {
+	// TODO: what should maxlen be?
+	b, err := ReadPrefix(d.r, 1<<32)
+	if err != nil {
+		panic(err)
+	}
+	return b
+}
+
+func (d *Decoder) decode(val reflect.Value) {
 	// check for UnmarshalSia interface first
 	if val.CanAddr() {
 		if u, ok := val.Addr().Interface().(SiaUnmarshaler); ok {
-			var dataLen int
-			dataLen, b = int(DecUint64(b[:8])), b[8:]
-			u.UnmarshalSia(b[:dataLen])
-			return dataLen + 8
+			u.UnmarshalSia(d.readPrefix())
+			return
 		}
 	}
 
 	switch val.Kind() {
 	case reflect.Ptr:
+		b := d.readN(1)
 		// nil pointer, nothing to decode
 		if b[0] == 0 {
-			return 1
+			return
 		}
 		// make sure we aren't decoding into nil
 		if val.IsNil() {
 			val.Set(reflect.New(val.Type().Elem()))
```

Comment:
nope, Marshal will panic if you try to marshal a bad type. Encode, however, can return an error, because it outputs to an `io.Writer` and the `Write` call can fail.